### DISCUSSION TITLE
A grain growth problem when only set high GB energy at GB based on `GBAnisotropy`

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#22254] [Q&A Modules: Phase field] [answered]
Dear MOOSE experts,

Recently, I used the [phase field model by Nele Moelans](https://doi.org/10.1103/PhysRevB.78.024113) and created class [GBsAnisotropyGG](https://github.com/PengWei97/luwu/blob/gNGsGG2022/grainGrowth/gNGs2022/error4/GBsAnisotropyGG.C) based on `GBAnisotropy`, with the aim of assigning high grain boundary energy 0.9 at grain boundaries and low grain boundary energy 0.5 inside the grains. This is to prepare for the next step of creating polycrystalline simulations that take into account the anisotropy of grain boundary energy due to misorientation between grains. However, by creating three bicrystal simulations, some problems were found, mainly in the morphology and dynamics of the grain boundary of the circular grain.

**My question is, from the phase field model modification point of view, is it correct to modify only the grain boundary energy tensor at the grain boundaries? Or perhaps it is a matter of numerical solution of the finite element method. Specifically, the derivative terms within the grains or in the bulk region are involved in the numerical solution of the entire domain and exert an influence on the evolution of the interface.**

In terms of specific code implementation, I created three bicrystal simulations, **Case I** with a grain boundary energy of 0.5 for the entire domain, including the grain boundaries and grain interiors, **Case II** with a grain boundary energy tensor of `GBAnisotropy` as shown below, and **Case III** with a grain boundary energy tensor of `GBAnisotropy` as shown below. And for Case III, it should be noted that the grain boundary energy is only assigned to 0.90 at the grain boundary and 0.50 elsewhere, which can be indirectly derived from the $\kappa$ cloud below.  For detailed code, input files and results, please see [the web page](https://github.com/PengWei97/luwu/tree/gNGsGG2022/grainGrowth/gNGs2022/error4).
![image](https://user-images.githubusercontent.com/56183775/192919880-7ad0e09f-aa59-42fa-9b7d-b6e91258383e.png)

Meanwhile, for the understanding of [Nele Moelans' multi-order parameter phase field model](https://doi.org/10.1103/PhysRevLett.101.025502), I think that the evolution of grain boundary is only related to the grain boundary energy and grain boundary mobility at the grain boundary, so it is reasonable to modify the grain boundary energy at the grain boundary only to reproduce the results of Case II perfectly.

However, the evolution of the output circular grain's area with time shows that the Case III curve is very jittery and close to the Case I curve. My expectation is that the kinetics of Case III and Case II are in perfect agreement. In addition, the shape of the grain boundary of the circular grain in Case III is also very rough and not smooth, which is caused by the numerical inhomogeneity.
![image](https://user-images.githubusercontent.com/56183775/192920386-c32cfd1f-56d4-4d1b-b39b-e8d8d8e866f8.png)

After that, the key code for the class `GBsAnisotropyGG` used in Case III is as follows.

```c++
  // get the grain boundary location based on the grainTracker
  const auto & op_to_grains = _grain_tracker.getVarToFeatureVector(_current_elem->id());

  std::vector<unsigned int> orderParameterIndex;

  for (MooseIndex(op_to_grains) op_index = 0; op_index < op_to_grains.size(); ++op_index)
  {
    auto grain_id = op_to_grains[op_index]; // grain id

    if (grain_id == FeatureFloodCount::invalid_id)
      continue;

    orderParameterIndex.push_back(op_index);
  }

  // if the quadrature point is located at the grain boundary, then set the grain boundary energy to 0.9
  if (orderParameterIndex.size() > 1)
    {
      for (unsigned int i = 0; i < _op_num -1; ++i)
        for (unsigned int j = i+1; j < _op_num; ++j)
          {
            _sigma[i][j] = 0.9;
            _sigma[j][i] = _sigma[i][j];
          }
    }
```
In general, I try to give the inside of the circular grain low grain boundary energy and the grain boundary region high grain boundary energy. This operation is intended to achieve the same simulation results in terms of grain boundary evolution and kinetic when both at the grain boundary and within the grain have high grain boundary energy. **The issue now is that I'm unsure if the appeal's inaccuracy was brought on by changing the phase field model or by a mistake in the numerical solution.**

Any suggestions or recommendations to fix these problems would be greatly appreciated.

Thanks

wei peng

**Comment by GiudGiud** [comment]
@laagesen @permcody

**Comment by laagesen** [comment, marked as answer]
In theory you could modify the grain boundary energy only at the GB itself, however from a practical standpoint I would not recommend this. In the grain interior where only one order parameter has its equilibrium value of 1 and all other order parameters are 0, there is no driving force for any change to any of the order parameters in the Allen-Cahn equations, so in theory you could set the material parameters to whatever you like and the order parameter still won't chance. However, the problem is, how do you define exactly where you transition from the grain boundary to grain interior happens? It looks like you are relying on the GrainTracker, but the problem is there is some numerical threshold you need to be above to go from a single grain to a GB region. The position where that happens is difficult to pinpoint exactly and dependent on solver tolerance among other factors. I think that is why you are seeing the interface not be circular. It is also often problematic to have large discontinuities in material properties such as you are introducing. So, I would not recommend to make the GB energy different away from GBs.

**Reply by PengWei97** [reply]
Thank you @laagesen for your reply. Base on your advice, I summarize it as two points,

1. Based on GrainTracker, and the tolerance of the solver and other factors, it is difficult to define where the transitions occure within the grain boundies and grains
2. large discontinuities in material properties, perhaps casusing problems such as numerical instability, I guess.

and I'm not sure if I understand it correctly.

**Reply by laagesen** [reply]
I think your summary captures everything right.

**Reply by PengWei97** [reply]
Good, please forgive me for still trying, the difference is that I want to modify the GB energy tensor of the whole domain in the polycrystalline model. The detailed idea is as follows,
1. at each quadrature point, obtain the corresponding vector of neighboring grain IDs for each grain ID, and later, according to the neighboring IDs, obtain the corresponding vector of the order parameter ID,
2. give the grain boundary energy tensor
           a. If only one order parameter is activated, get the nearest neighboring grain ID according to the coordinates of that point and the sequence of center coordinates of the neighboring grains, and then assign the GB energy tensor,
           b. If at least two order parameter are activated, then the GB energy tensor is assigned, as described above

I think such a strategy can be used to artificially assign the GB energy tensor to different grains (both within the grain and at the GB) as a whole. It can also solve the two problems mentioned above. I can currently output the adjacent grain ID vector, and the next step is to get the index of the order parameter based on the grain ID number. I am not too confident that this will achieve prefetching, and I would appreciate any suggestions you have for this strategy of mine and guidance to accomplish this goal.

By the way, my ultimate goal is to combine EBSD and construct a grain growth model considering GB energy aniostropy depend on misorientation.

**Comment by Hejie-Guo** [comment]
Hello, dear Wei, I am trying to model the anisotropic grain growth, could you please give some advice. 

I am trying to model elongate temperature-dependent grain growth behavior from experimental obervations. The anisotropic grain growth behavior is similar as the following figures:
![捕获11](https://github.com/idaholab/moose/assets/61571563/560e96ef-8233-49eb-8e02-1af93eb26e07)


I simply modify some moose examples and modeling the temperature-dependent behavior, but did not quite understand how to model the anisotropic grain growth. Here is the result i can get using moose, i just want to anisotropic modeling the shape as the black line.
![捕获](https://github.com/idaholab/moose/assets/61571563/ff1e8f17-ed63-4fa9-b8d2-525d5c8244ea)
